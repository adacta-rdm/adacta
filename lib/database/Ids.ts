import type { Opaque } from "type-fest";

export type IDeviceDefinitionId = Opaque<string, "DEVICE_DEFINITION">;
export type IDeviceId = Opaque<string, "DEVICE">;
export type IIdPoolId = Opaque<string, "ID_POOL">;
export type IImportPresetId = Opaque<string, "IMPORT_PRESET">;
export type INameCompositionId = Opaque<string, "NAME_COMPOSITION">;
export type INameCompositionVariableId = Opaque<string, "NAME_COMPOSITION_VARIABLE">;
export type INoteId = Opaque<string, "NOTE">;
export type IProjectId = Opaque<string, "PROJECT">;
export type IPropertyId = Opaque<string, "PROPERTY">;
export type IRepositorySettingId = Opaque<string, "REPOSITORY_SETTING">;
export type IResourceId = Opaque<string, "RESOURCE">;
export type ISampleId = Opaque<string, "SAMPLE">;
export type ISampleRelationId = Opaque<string, "SAMPLE_RELATION">;
export type ITransformationId = Opaque<string, "TRANSFORMATION">;
export type IUserId = Opaque<string, "USER">;

/**
 * Maps the GraphQL __typename to their respective ID types.
 * Note: This is only required for types that are not part of CouchDB or types that are named
 * differently than the "type" field of the CouchDB document.
 */
export interface IIdTypeMap {
	// Register types from auth-server (prisma)
	User: IUserId;
	RemoteRepo: IRepositorySettingId;

	// Register alias types from auth-server
	AuthenticatedUser: IUserId;

	// Register types from repo-server (MikroORM)
	Note: INoteId;

	// Register alias types from repo-server
	ResourceGeneric: IResourceId;
	ResourceImage: IResourceId;
	ResourceImportPreset: IResourceId;
	ResourceTabularData: IResourceId;

	// Register MikroORM types
	NameComposition: INameCompositionId;
	NameCompositionVariable: INameCompositionVariableId;
	NameCompositionVariableStatic: INameCompositionVariableId;
	NameCompositionVariableVariable: INameCompositionVariableId;
}
