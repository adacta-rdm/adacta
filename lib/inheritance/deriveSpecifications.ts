import lodash from "lodash";

import type { IDeviceDefinitionId, ISampleId } from "~/lib/database/Ids";

interface ITraversalMetadata {
	/**
	 * Length of shortest path from initial Device/DeviceDefinition to the Specification
	 */
	level: number;
}

export interface ISpecificationTraversalResult extends ITraversalMetadata {
	name: string;
	value: string;
	source?: { id: string; name: string };
	skipped?: boolean; // TODO: Add into different interface?
}

/**
 * Contains a reference to the definition and metadata about the traversal
 * Used
 */
export interface IDeviceDefinitionTraversalResult extends ITraversalMetadata {
	definition: IDeviceDefinitionId;
}

export interface ISampleTraversalResult extends ITraversalMetadata {
	sample: ISampleId;
}

export function deriveSpecifications(
	specifications: ISpecificationTraversalResult[],
	includeSkipped = false
): ISpecificationTraversalResult[] {
	const grouped = Object.values(lodash.groupBy(specifications, (s) => s.name));

	return grouped
		.flatMap((group) => {
			if (group.length == 1) {
				return selectElement(group, includeSkipped);
			}

			// Sort ascending by level
			group.sort((a, b) => a.level - b.level);

			return selectElement(group, includeSkipped);
		})
		.flat();
}

function selectElement(elements: ISpecificationTraversalResult[], includeSkipped: boolean) {
	if (!includeSkipped) {
		return elements[0];
	}

	const [first, ...skipped] = elements;
	return [first, skipped.map((s) => ({ ...s, skipped: true }))];
}
