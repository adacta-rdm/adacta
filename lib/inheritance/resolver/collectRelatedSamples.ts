import assert from "node:assert";

import { eq } from "drizzle-orm";

import type { ISampleId } from "../../database/Ids";
import type { ISampleTraversalResult } from "../deriveSpecifications";

import type { EntityLoader } from "~/apps/repo-server/src/services/EntityLoader";
import type { DrizzleSchema } from "~/drizzle/DrizzleSchema";

export async function collectRelatedSamples(
	startId: ISampleId,
	el: EntityLoader,
	schema: DrizzleSchema,
	level = 0
): Promise<ISampleTraversalResult[]> {
	const { SampleToSample } = schema;
	const samples: ISampleTraversalResult[] = [];
	samples.push({ sample: startId, level });

	const parents = await el.find(SampleToSample, (t) => eq(t.sample2, startId));
	assert(parents.length <= 1, "Sample must have one or no parent");

	// If there is no parent we are done as it is the root of the tree
	if (parents.length === 0) {
		return samples;
	}

	const parent = parents[0];

	samples.push(...(await collectRelatedSamples(parent.sample1, el, schema, ++level)));

	return samples;
}
