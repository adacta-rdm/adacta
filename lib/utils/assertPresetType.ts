import { assertICSVPreset, assertIGamryPreset } from "@/tsrc/lib/interface/IImportWizardPreset";
import { ImportPresetType } from "~/drizzle/schema/repo.ImportPreset";
import type { ICSVPreset, IGamryPreset } from "~/lib/interface/IImportWizardPreset";

export function assertPresetType<T extends ImportPresetType>(
	type: T,
	preset: ICSVPreset | IGamryPreset
): asserts preset is T extends ImportPresetType.CSV ? ICSVPreset : IGamryPreset {
	if (type === ImportPresetType.CSV) {
		assertICSVPreset(preset);
	} else if (type === ImportPresetType.GAMRY) {
		assertIGamryPreset(preset);
	} else {
		throw new Error(`Invalid preset type '${type}'`);
	}
}
