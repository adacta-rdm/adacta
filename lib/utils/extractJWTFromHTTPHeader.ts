import { StatusCodes } from "http-status-codes";
import { verify } from "jsonwebtoken";
import { isObject } from "lodash";

import { readEnvVar } from "./readEnvVar";
import { HTTPAuthenticationError } from "../errors/HTTPAuthenticationError";
import type { IAuthServerJWTParams } from "../interface/IAuthServerJWTParams";

// See https://www.npmjs.com/package/jsonwebtoken#algorithms-supported for a list of possible algos.

export function extractJWTFromHTTPHeader(
	authorizationHeader: string | undefined,
	AUTH_SERVER_JWT_PUBLIC_KEY = readEnvVar("AUTH_SERVER_JWT_PUBLIC_KEY")
): IAuthServerJWTParams {
	if (!authorizationHeader) throw new HTTPAuthenticationError(StatusCodes.UNAUTHORIZED);
	const [keyword, token] = authorizationHeader.split(" ");
	if (keyword !== "Bearer") throw new HTTPAuthenticationError(StatusCodes.BAD_REQUEST);

	const payload = verify(token, AUTH_SERVER_JWT_PUBLIC_KEY, {
		algorithms: ["RS256"],
	});

	if (!isObject(payload) || !("userId" in payload)) {
		throw new HTTPAuthenticationError(StatusCodes.BAD_REQUEST);
	}

	// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment,@typescript-eslint/no-unsafe-member-access
	return { userId: (payload as any).userId };
}
