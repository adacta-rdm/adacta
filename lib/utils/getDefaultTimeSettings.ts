import type { ITimeSettings } from "../interface/ITimeSettings";

export function getDefaultTimeSettings(): ITimeSettings {
	return {
		timeStyle: "short",
		dateStyle: "short",
		locale: "de-DE",
	};
}
