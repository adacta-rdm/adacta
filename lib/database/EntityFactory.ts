import type { IUserId } from "./Ids";

import { AuthenticatedUserInfo } from "~/apps/repo-server/src/graphql/AuthenticatedUserInfo";
import { entityId } from "~/apps/repo-server/src/utils/entityId";
import type { DrizzleEntity, DrizzleEntityNameId, Metadata } from "~/drizzle/DrizzleSchema";
import { entityByTypeId } from "~/drizzle/DrizzleSchema";
import { Service } from "~/lib/serviceContainer/ServiceContainer";

@Service(AuthenticatedUserInfo)
export class EntityFactory {
	static id<T extends DrizzleEntityNameId>(e: T): DrizzleEntity<T>["id"] {
		const entry = Object.entries(entityByTypeId).find(([, v]) => v === e);
		if (!entry) throw new Error(`No entity with typeid ${e}`);
		return entityId(entry[0]) as DrizzleEntity<T>["id"];
	}

	static create<T extends "User", TProps extends EntityProps<"User">>(
		entityName: T,
		props: TProps
	): TProps & { id: DrizzleEntity<T>["id"] };
	static create<T extends DrizzleEntityNameId, TProps extends EntityProps<T>>(
		entityName: T,
		props: TProps,
		creatorId: IUserId
	): TProps & Metadata & { id: DrizzleEntity<T>["id"] };
	static create<T extends DrizzleEntityNameId, TProps extends EntityProps<T>>(
		entityName: T,
		props: TProps,
		creatorId?: IUserId
	) {
		const o = {
			...props,
			id: EntityFactory.id(entityName),
		};

		if (entityName === "User") return o;

		if (!creatorId) {
			throw new Error(`Missing required argument 'creatorId' for entity ${entityName}`);
		}

		return Object.assign(o, EntityFactory.metadata(creatorId));
	}

	constructor({ userId }: AuthenticatedUserInfo) {
		this.userId = userId;
	}

	create<T extends DrizzleEntityNameId, TProps extends EntityProps<T>>(
		entityName: T,
		props: TProps
	) {
		return EntityFactory.create(entityName, props, this.userId);
	}

	private userId: IUserId;

	private static metadata(metadataCreatorId: IUserId) {
		return {
			metadataCreatorId,
			metadataCreationTimestamp: new Date(),
			metadataDeletedAt: null,
		} satisfies Metadata;
	}
}

type EntityProps<T extends DrizzleEntityNameId> = Omit<
	DrizzleEntity<T>,
	// Remove the id/metadata fields as these are populated by the factory
	"metadataCreatorId" | "metadataCreationTimestamp" | "metadataDeletedAt" | "id"
>;
