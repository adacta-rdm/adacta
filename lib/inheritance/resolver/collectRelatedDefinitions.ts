import assert from "assert";

import { and, eq, gt } from "drizzle-orm";

import type { IDeviceDefinitionTraversalResult } from "../deriveSpecifications";

import type { EntityLoader } from "~/apps/repo-server/src/services/EntityLoader";
import { decodeEntityId } from "~/apps/repo-server/src/utils/decodeEntityId";
import type { DrizzleSchema } from "~/drizzle/DrizzleSchema";
import type { IDeviceDefinitionId, IDeviceId } from "~/lib/database/Ids";

/**
 * Starts from a Device or Device definition and collects all related device definitions.
 * For Devices, the Device definition of this device are collected.
 * For DeviceDefinitions all parent device definitions are collected.
 */
export async function collectRelatedDefinitions(
	startId: IDeviceId | IDeviceDefinitionId,
	ctx: { el: EntityLoader; schema: DrizzleSchema },
	level = 0
): Promise<IDeviceDefinitionTraversalResult[]> {
	const {
		el,
		schema: { Device, DeviceDefinition, DeviceDefinitionPaths },
	} = ctx;

	const definitions: IDeviceDefinitionTraversalResult[] = [];

	const startDocType = decodeEntityId(startId);
	assert(startDocType === "Device" || startDocType === "DeviceDefinition"); // TODO: Change signature to include type? I think we usually know the type...

	if (startDocType === "Device") {
		const startDoc = await el.one(Device, startId);
		definitions.push(...(await collectRelatedDefinitions(startDoc.definitionId, ctx, ++level)));

		const results = await collectRelatedDefinitions(startDoc.definitionId, ctx, level + 1);

		// Continue traversal on DeviceDefinitions
		definitions.push(...results);
	} else if (startDocType === "DeviceDefinition") {
		const startDoc = await el.one(DeviceDefinition, startId);
		definitions.push({ definition: startDoc.id, level });

		const parents = await el.drizzle
			.select({ id: DeviceDefinitionPaths.ancestorId, depth: DeviceDefinitionPaths.depth })
			.from(DeviceDefinitionPaths)
			.where(
				and(eq(DeviceDefinitionPaths.descendantId, startDoc.id), gt(DeviceDefinitionPaths.depth, 0))
			);

		definitions.push(
			...parents.map(({ id, depth }) => ({
				definition: id,
				level: depth + level,
			}))
		);
	}

	return definitions;
}
