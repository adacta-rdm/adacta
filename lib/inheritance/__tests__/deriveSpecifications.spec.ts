import { eq } from "drizzle-orm";
import { beforeEach, describe, expect, test } from "vitest";

import { EntityFactory } from "../../database/EntityFactory";
import type { IUserId } from "../../database/Ids";
import type { IDeviceDefinitionTraversalResult } from "../deriveSpecifications";
import { deriveSpecifications } from "../deriveSpecifications";
import { collectRelatedDefinitions } from "../resolver/collectRelatedDefinitions";

import type { EntityLoader } from "~/apps/repo-server/src/services/EntityLoader";
import { closureTableSetParent } from "~/apps/repo-server/src/utils/closureTable/closureTableTree";
import { setSpecifications } from "~/apps/repo-server/src/utils/setSpecifications";
import { createTestDb } from "~/apps/repo-server/testUtils";
import type { DrizzleEntity, DrizzleSchema } from "~/drizzle/DrizzleSchema";

describe("Specifications inheritance", () => {
	let el: EntityLoader;
	let schema: DrizzleSchema;

	let userId: IUserId;

	let bronkhorstDevices: DrizzleEntity<"DeviceDefinition">;
	let mfcDevices: DrizzleEntity<"DeviceDefinition">;
	let mfc: DrizzleEntity<"Device">;

	const convertDerivedDefinitions = async (d: IDeviceDefinitionTraversalResult[]) => {
		return (
			await Promise.all(
				d.map((d) =>
					(async () => {
						return (
							await el.drizzle
								.select()
								.from(schema.DeviceDefinitionSpecification)
								.where(eq(schema.DeviceDefinitionSpecification.ownerId, d.definition))
						).map((s) => ({
							level: d.level,
							name: s.name,
							value: s.value,
						}));
					})()
				)
			)
		).flat();
	};

	beforeEach(async () => {
		const testDb = await createTestDb();
		el = testDb.el;
		schema = testDb.schema;

		userId = testDb.user.id;

		bronkhorstDevices = EntityFactory.create(
			"DeviceDefinition",
			{
				name: "Bronkhorst Devices",
				imageResourceIds: [],
				acceptsUnit: [],
			},
			userId
		);
		await el.insert(schema.DeviceDefinition, bronkhorstDevices);
		await setSpecifications(
			bronkhorstDevices.id,
			[{ name: "manufacture", value: "Bronkhorst" }],
			schema.DeviceDefinitionSpecification,
			el
		);

		// MFCs
		mfcDevices = EntityFactory.create(
			"DeviceDefinition",
			{
				name: "MFCs",
				imageResourceIds: [],
				acceptsUnit: [],
			},
			userId
		);
		await el.insert(schema.DeviceDefinition, mfcDevices);
		await closureTableSetParent(
			mfcDevices.id,
			bronkhorstDevices.id,
			el,
			schema.DeviceDefinitionPaths
		);
		await setSpecifications(
			mfcDevices.id,
			[{ name: "description", value: "Gas mass flow controllers for flow rates" }],
			schema.DeviceDefinitionSpecification,
			el
		);

		mfc = EntityFactory.create(
			"Device",
			{
				name: "Test-MFC",
				definitionId: mfcDevices.id,
				imageResourceIds: [],
				setupDescription: [],
			},
			userId
		);

		await el.insert(schema.Device, mfc);
		await setSpecifications(
			mfc.id,
			[{ name: "serialNo", value: "123-123" }],
			schema.DeviceSpecification,
			el
		);
	});

	test("Basic", async () => {
		expect(
			deriveSpecifications(
				await convertDerivedDefinitions(await collectRelatedDefinitions(mfc.id, { el, schema }))
			).map((s) => ({
				name: s.name,
				value: s.value,
			}))
		).toMatchSnapshot();
	});

	test("More-Specific values should be used", async () => {
		// Insert second description (both bronkhorstDevices and mfcDevices now have a description)
		// We want the mfcDevices description to win since it is more specific (closer to the device
		// in the tree)
		await el.drizzle
			.insert(schema.DeviceDefinitionSpecification)
			.values({ ownerId: bronkhorstDevices.id, name: "description", value: "Bronkhorst-Devices" });
		await el.update(schema.DeviceDefinition, bronkhorstDevices.id, bronkhorstDevices);

		expect(
			deriveSpecifications(
				await convertDerivedDefinitions(await collectRelatedDefinitions(mfc.id, { el, schema }))
			).map((s) => ({
				name: s.name,
				value: s.value,
			}))
		).toMatchSnapshot();
	});
});
