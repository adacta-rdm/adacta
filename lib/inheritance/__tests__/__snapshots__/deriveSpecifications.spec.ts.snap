// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Specifications inheritance > Basic 1`] = `
[
  {
    "name": "description",
    "value": "Gas mass flow controllers for flow rates",
  },
  {
    "name": "manufacture",
    "value": "Bronkhorst",
  },
]
`;

exports[`Specifications inheritance > Inherit from multiple devices 1`] = `
[
  {
    "name": "color",
    "value": "yellow",
  },
  {
    "name": "manufacture",
    "value": "Bronkhorst",
  },
  {
    "name": "description",
    "value": "Bronkhorst-Devices",
  },
]
`;

exports[`Specifications inheritance > Inherit from multiple devices 2`] = `
[
  {
    "name": "color",
    "value": "yellow",
  },
  {
    "name": "description",
    "value": "Gas mass flow controllers for flow rates",
  },
  {
    "name": "manufacture",
    "value": "Bronkhorst",
  },
]
`;

exports[`Specifications inheritance > More-Specific values should be used 1`] = `
[
  {
    "name": "description",
    "value": "Gas mass flow controllers for flow rates",
  },
  {
    "name": "manufacture",
    "value": "Bronkhorst",
  },
]
`;

exports[`closureTable > Delete Leaf 1`] = `
[
  {
    "ancestor": "0",
    "depth": 1,
    "descendant": "1",
  },
  {
    "ancestor": "0",
    "depth": 2,
    "descendant": "2",
  },
  {
    "ancestor": "1",
    "depth": 1,
    "descendant": "2",
  },
]
`;

exports[`closureTable > Delete subtree 1`] = `
[
  {
    "ancestor": "0",
    "depth": 1,
    "descendant": "1",
  },
]
`;

exports[`closureTable > Inserts into closure table 1`] = `
[
  {
    "ancestor": "0",
    "depth": 1,
    "descendant": "1",
  },
]
`;

exports[`closureTable > Inserts into closure table 2`] = `
[
  {
    "ancestor": "0",
    "depth": 1,
    "descendant": "1",
  },
  {
    "ancestor": "0",
    "depth": 2,
    "descendant": "2",
  },
  {
    "ancestor": "1",
    "depth": 1,
    "descendant": "2",
  },
]
`;

exports[`closureTable > Inserts into closure table 3`] = `
[
  {
    "ancestor": "0",
    "depth": 1,
    "descendant": "1",
  },
  {
    "ancestor": "0",
    "depth": 2,
    "descendant": "2",
  },
  {
    "ancestor": "0",
    "depth": 3,
    "descendant": "3",
  },
  {
    "ancestor": "0",
    "depth": 3,
    "descendant": "4",
  },
  {
    "ancestor": "1",
    "depth": 1,
    "descendant": "2",
  },
  {
    "ancestor": "1",
    "depth": 2,
    "descendant": "3",
  },
  {
    "ancestor": "1",
    "depth": 2,
    "descendant": "4",
  },
  {
    "ancestor": "2",
    "depth": 1,
    "descendant": "3",
  },
  {
    "ancestor": "2",
    "depth": 1,
    "descendant": "4",
  },
]
`;

exports[`closureTable > Move subtree 1`] = `
[
  {
    "ancestor": "0",
    "depth": 1,
    "descendant": "1",
  },
  {
    "ancestor": "0",
    "depth": 2,
    "descendant": "2",
  },
  {
    "ancestor": "0",
    "depth": 3,
    "descendant": "3",
  },
  {
    "ancestor": "0",
    "depth": 4,
    "descendant": "4",
  },
  {
    "ancestor": "0",
    "depth": 5,
    "descendant": "5",
  },
  {
    "ancestor": "1",
    "depth": 1,
    "descendant": "2",
  },
  {
    "ancestor": "1",
    "depth": 2,
    "descendant": "3",
  },
  {
    "ancestor": "1",
    "depth": 3,
    "descendant": "4",
  },
  {
    "ancestor": "1",
    "depth": 4,
    "descendant": "5",
  },
  {
    "ancestor": "2",
    "depth": 1,
    "descendant": "3",
  },
  {
    "ancestor": "2",
    "depth": 2,
    "descendant": "4",
  },
  {
    "ancestor": "2",
    "depth": 3,
    "descendant": "5",
  },
  {
    "ancestor": "3",
    "depth": 1,
    "descendant": "4",
  },
  {
    "ancestor": "3",
    "depth": 2,
    "descendant": "5",
  },
  {
    "ancestor": "4",
    "depth": 1,
    "descendant": "5",
  },
]
`;
